src/calendarworker.cpp

```diff
@@ -752,23 +752,19 @@ CalendarWorker::eventOccurrences(const QList<CalendarData::Range> &ranges) const
 
 QHash<QDate, QStringList>
 CalendarWorker::dailyEventOccurrences(const QList<CalendarData::Range> &ranges,
-                                      const QList<CalendarData::EventOccurrence> &occurrences)
+                                      const QList<CalendarData::EventOccurrence> &occurrences) const
 {
     QHash<QDate, QStringList> occurrenceHash;
-    foreach (const CalendarData::Range &range, ranges) {
-        QDate start = range.first;
-        while (start <= range.second) {
-            foreach (const CalendarData::EventOccurrence &eo, occurrences) {
-                const QDateTime startDt(start); // To be replaced later by start.startOfDay()
-                // On all day events the end time is inclusive, otherwise not
-                if ((eo.eventAllDay && eo.startTime.date() <= start
-                     && eo.endTime.date() >= start)
-                    || (!eo.eventAllDay && eo.startTime < startDt.addDays(1)
-                        && eo.endTime >= startDt)) {
-                    occurrenceHash[start].append(eo.getId());
-                }
+    for (const CalendarData::EventOccurrence &eo : occurrences) {
+        const QDate st = eo.eventAllDay ? eo.startTime.date() : eo.startTime.toLocalTime().date();
+        const QDate ed = eo.eventAllDay ? eo.endTime.date() : eo.endTime.toLocalTime().date();
```

Comment:
Not sure I'm following. Earlier there was the allday inclusivity check but how is toLocalTime() related ot that or why such is needed?

Reply:
When I exchange the `for` loops the previous inner test was :
```
   if ((eo.eventAllDay && eo.startTime.date() <= start
                     && eo.endTime.date() >= start)
                    || (!eo.eventAllDay && eo.startTime < startDt.addDays(1)
                        && eo.endTime >= startDt)) {
                    occurrenceHash[start].append(eo.getId());
                }
```
with `start` being the date when the occurrence may happen or not.

If I read this `if` condition properly, it says:
* in case of all day events, `start` is taken into account if the occurrence first day is at `start` or before and also if the occurrence last day is at `start` or later.
* in case of non all day events, `start` is taken into account if the start time in the local time zone is during `start` date or earlier, and if the end time in the local time zone is later than the start of day of `start`.

I reverse the `for` loop order so the inner one is now iterating on dates the occurrence is spanning over, following the same convention as the `if` before, namely :
* for all day events, the start date is taken as the `date()` of the startTime, without paying attention to the time zone, like before, and spanning up to the `date()` of endTime. I think doing like that, I'm respecting the `<=` and `>=` that were previously there.
* for non all day event, the date must be taken in the local time zone, thus the conversion.

The agendamodel test supposed to check this is actually working with the patch. I was not sure myself of not having done any mistake when simplifying the routine like that, but it seems ok. That being said, an addition look and explanation is of course not a wasted time. Tell me if you agree with my reasoning.

Reply:
Thinking more about this, in particular about the mention that non all day are end exclusive, I now believe that the previous `if` was a bit erroneous. Indeed for the all day events, the test on end time is done with a `>=`, but it is also done like that for non all day event. I think it should have finished with `&& eo.endTime > startDt))`.

I've just amended the commit to add a `.addSecs(-1)` to the generation of the date in the non all day case. I've readded the comment also to make it clear.

With it, I can create an event from 10pm to 00:00 the day after, and having it displayed only on the correct day and not having a 00:00 to 00:00 occurrence also the day after in the calendar.

Thanks a lot for noticing this and warning me.

Reply:
Yea, the the inclusive / exclusive looks better now. Resulting times from the event being always inclusive.

The remaining detail I'm pondering with this toLocalTime difference is all-day events with timezone set to non-current timezone. Those are tricky, likely not what the user expects and that's why the Calendar app creates all-day events without timezone. But remotely synced event could contains such. E.g. I'm in Finland and there's a 1 one day all-day event with some US timezone. That should end up shown on two days, right?

Reply:
The problem of time zone for full day events is a tricky one and I think I'm keeping the previous behaviour here. Namely, we convert the occurrence date time (start or end) into a date without paying attention to the time zone it was defined into (see previous code like `eo.startTime.date() <= start`, where `start` is a date itself.

We discussed this a long time ago many time on IRC with @chriadam and we arrive to the conlusion that :
* case DTSTART is a DATE, like in DTSTART:20220120 cannot has a time zone definition (see https://datatracker.ietf.org/doc/html/rfc5545#section-3.8.3.1). It is always resolving as a full day whatever the time zone. For instance, my example will start at 20220119T180000Z when you are in Ho Chi Minh city (UTC+06).
* case DTSTART is a DATE-TIME, without time zone and starting at 00:00 local time, like DTSTART:20220120T000000 (duration is supposed to be a multiple of a day). mKCal is flagging such as full day events (see sqliteformat.cpp#1216).
* case DTSTART is a DATE-TIME with a time zone and starting at 00:00 in this time zone, like DTSTART;TZID=Asia/Ho_Chi_Minh:20220120T000000. In that case, we are considering this event as _not_ being full day, even in Viet Nam time zone. It is just an event that happens from midnight. It is not flagged as full day by mKCal.

The two first cases are internally stored by KCalendarCore (and restored by mKCal) as an event `e` with `e.fullDay() == true` and `e.dtStart().time() == 00:00` and `e.dtStart().spec() == LocalTime`. Thus, with this patch (and also without) they are taking the fullDay branch in the condition and the date is taken without paying attention to the time zone (there is not any anyway, spec being LocalTime).

The last case is stored internally by KCalendarCore (and restored by mKCal) as an event `e` with `e.fullDay() == false` and `e.dtStart().time() == 00:00` and `e.dtStart().spec() == TimeZone`. Then, it will enter the branch for non full day and the date will be extracted _after_ a conversion to the local zone of the device (and may thus be visible during two days on the device if the local time zone is different from the one of the event).

I think that this is thus respecting the three cases cited above.

Reply:
Fair enough. I could consider it a bug in mkcal if it restores an event different than what was saved to the database, e.g. losing the allday flag if there was one earlier. But that could be a different thing and maybe not the highest priority thing to revisit now.